### PR TITLE
Fix index permissions for negative lookahead and negated regex index patterns

### PR DESCRIPTION
**Pull request by andy840314**
##  opensearch-security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Bug fix
 
 
 2. __Github Issue # or road-map entry, if available:__

 3. __Description of changes:__
All calls to Resolved.getAllIndices() and Resolved.getAliases() must be guarded by a !requestedResolved.isLocalAll().

 4. __Why these changes are required?__ 
 
The conditions necessary for triggering the bug are:

- security plugin is configured with do_not_fail_on_forbidden: false
- The role definitions use an index pattern with negation. This can be either negative lookahead (like /^(?!t.*).*/) or a negated character class (like [^x-y].*).
- A user makes a search request on all indices using the * pattern.

In this case, the user is granted access to all indices, even though the negation should restrict access to certain indexes (in the example with the index pattern /^(?!t.*).*/ the user should have no access for indices starting with a t).

This issue can only happen if all three conditions are true. So, there is no security issue if do_not_fail_on_forbidden is true, or if there are no index patterns with negation, or if the search operates on concrete index names or wildcards other than * (i.e., /_search/a* is not affected).

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
According to https://github.com/opensearch-project/security/pull/1335#issuecomment-901256219, this is now blocking https://github.com/opensearch-project/security/pull/1335 which is blocking https://github.com/opensearch-project/security/pull/1409 which is blocking https://github.com/opensearch-project/opensearch-build/issues/179.

File: src/main/java/org/opensearch/security/configuration/PrivilegesInterceptorImpl.java [hunk 1]
```diff
@@ -341,11 +341,15 @@ private String toUserIndexName(final String originalDashboardsIndex, final Strin
     }
 
     private static boolean resolveToDashboardsIndexOrAlias(final Resolved requestedResolved, final String dashboardsIndexName) {
-        final Set<String> allIndices = requestedResolved.getAllIndices();
-        if (allIndices.size() == 1 && allIndices.iterator().next().equals(dashboardsIndexName)) {
-            return true;
+        if (!requestedResolved.isLocalAll()) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/configuration/PrivilegesInterceptorImpl.java:344]:
```suggestion
        if (requestedResolved.isLocalAll()) {
            return false;
        }    
```

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java [hunk 2]
```diff
@@ -420,8 +420,11 @@ private boolean checkImmutableIndices(Object request, ActionListener listener) {
 
     private boolean isRequestIndexImmutable(Object request) {
         final IndexResolverReplacer.Resolved resolved = indexResolverReplacer.resolveRequest(request);
-        final Set<String> allIndices = resolved.getAllIndices();
-
-        return immutableIndicesMatcher.matchAny(allIndices);
+        if (resolved.isLocalAll()) {
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/filter/SecurityFilter.java:423]:
```suggestion
        if (resolved.isLocalAll()) {
            return true;
        }     
```

File: src/main/java/org/opensearch/security/privileges/DlsFlsEvaluator.java [hunk 3]
```diff
@@ -100,8 +100,8 @@ public PrivilegesEvaluatorResponse evaluate(final ActionRequest request, final C
             }
 
             presponse.maskedFields = maskedFieldsMap.entrySet().stream()
-                .filter(requestedResolved.getAllIndices().isEmpty() ?
-                    entry -> true : entry -> WildcardMatcher.from(entry.getKey()).matchAny(requestedResolved.getAllIndices()))
+                .filter(!requestedResolved.isLocalAll() && !requestedResolved.getAllIndices().isEmpty() ?
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/privileges/DlsFlsEvaluator.java:103]: convert two `NOT` to `OR`.

File: src/main/java/org/opensearch/security/privileges/DlsFlsEvaluator.java [hunk 4]
```diff
@@ -134,8 +134,8 @@ public PrivilegesEvaluatorResponse evaluate(final ActionRequest request, final C
             }
 
             presponse.queries = dlsQueries.entrySet().stream()
-                .filter(requestedResolved.getAllIndices().isEmpty() ?
-                        entry -> true : entry -> WildcardMatcher.from(entry.getKey()).matchAny(requestedResolved.getAllIndices()))
+                .filter(!requestedResolved.isLocalAll() && !requestedResolved.getAllIndices().isEmpty() ?
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/privileges/DlsFlsEvaluator.java:137]: Same as the other one.

File: src/main/java/org/opensearch/security/privileges/DlsFlsEvaluator.java [hunk 5]
```diff
@@ -165,8 +165,8 @@ public PrivilegesEvaluatorResponse evaluate(final ActionRequest request, final C
             }
 
             presponse.allowedFlsFields = flsFields.entrySet().stream()
-                .filter(requestedResolved.getAllIndices().isEmpty() ?
-                        entry -> true : entry -> WildcardMatcher.from(entry.getKey()).matchAny(requestedResolved.getAllIndices()))
+                .filter(!requestedResolved.isLocalAll() && !requestedResolved.getAllIndices().isEmpty() ?
```
Comment (vrozov) [inline src/main/java/org/opensearch/security/privileges/DlsFlsEvaluator.java:168]: Same as the other one.

File: src/test/resources/roles_index_patterns.yml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+negated_regex_role:
+  description: "Role with negated regex index pattern"
+  cluster_permissions:
+    - "cluster_composite_ops"
+  index_permissions:
+    - index_patterns:
+        - "/^[a-z].*/"
+      allowed_actions:
+        - "read"
```
Comment (cliu123) [inline src/test/resources/roles_index_patterns.yml:24]: Add an empty line.

**Review by cliu123 — APPROVED** [review]

File: src/main/java/org/opensearch/security/privileges/ProtectedIndexAccessEvaluator.java [hunk 7]
```diff
@@ -83,8 +84,8 @@ public PrivilegesEvaluatorResponse evaluate(final ActionRequest request, final T
             presponse.allowed = false;
             return presponse.markComplete();
         }
-        if((indexMatcher.matchAny(requestedResolved.getAllIndices())
-                || requestedResolved.isLocalAll())
+        if((requestedResolved.isLocalAll()
```
Comment (palashhedau) [inline src/main/java/org/opensearch/security/privileges/ProtectedIndexAccessEvaluator.java:87]: any reason to change the sequence?

**Review by palashhedau — APPROVED** [review]

File: src/main/java/org/opensearch/security/privileges/ProtectedIndexAccessEvaluator.java (hunk 7 above)
Comment (andy840314) [reply src/main/java/org/opensearch/security/privileges/ProtectedIndexAccessEvaluator.java:87]: isLocalAll() need to be called first before getAllIndices()

**Pull request merged**